mergin/client_pull.py

```diff
@@ -0,0 +1,489 @@
+        return
+    
+    # TODO: make download_blocking / save_to_file cancellable so that we can cancel as soon as possible
+
+    item.download_blocking(mc, project_path)
+    job.transferred_size += item.size
+    #print(threading.current_thread(), "downloading", item.file_path, "finished")
+
+
+def download_project_async(mc, project_path, directory):
+    """
+    Starts project download in background and returns handle to the pending project download.
+    Using that object it is possible to watch progress or cancel the ongoing work.
+    
+    """
+
+    if os.path.exists(directory):
+        raise Exception("Project directory already exists")
+    os.makedirs(directory)
+    mp = MerginProject(directory)
+
+    project_info = mc.project_info(project_path)
+    version = project_info['version'] if project_info['version'] else 'v0'
+
+    # prepare download
+    update_tasks = []  # stuff to do at the end of download
+    for file in project_info['files']:
+        file['version'] = version
+        items = _download_items(file, directory)
+        update_tasks.append(UpdateTask(file['path'], items))
+
+    # make a single list of items to download
+    total_size = 0
+    download_list = []
+    for task in update_tasks:
+        download_list.extend(task.download_queue_items)
+        for item in task.download_queue_items:
+            total_size += item.size
+    
+    job = DownloadJob(project_path, total_size, version, update_tasks, download_list, directory, mp, project_info)
+    
+    # start download
+    job.executor = concurrent.futures.ThreadPoolExecutor(max_workers=4)
+    job.futures = []
+    for item in download_list:
+        future = job.executor.submit(_do_download, item, mc, project_path, job)
+        job.futures.append(future)
+
+    return job
+
+
+def download_project_wait(job):
+    """ blocks until all download tasks are finished """
+    
+    concurrent.futures.wait(job.futures)
+
+    # handling of exceptions
+    for future in job.futures:
+        if future.exception() is not None:
+            raise future.exception()
+
+
+def download_project_is_running(job):
+    """ Returns true/false depending on whether we have some pending downloads """
+    for future in job.futures:
+        if future.running():
+            return True
+    return False
+
+
+def download_project_finalize(job):
+    """
+    To be called when download in the background is finished and we need to do the finalization (merge chunks etc.)
+    
+    This probably should not be called from a worker thread (e.g. directly from a handler when download is complete)
+    """
+
+    job.executor.shutdown(wait=True)
+    
+    assert job.transferred_size == job.total_size
+
+    for task in job.update_tasks:
+        
+        # right now only copy tasks...
+        task.apply(job.directory, job.mp)
+    
+    # final update of project metadata
+    # TODO: why not exact copy of project info JSON ?
```

Comment:
I think we need to revisit what is server sending but I guess we just do not store irrelevant information, I agree we should simplify this. Also not sure if this function should update metadata since we are just referencing to mp or move it elsewhere. But we can revisit in future.

Reply:
yeah... I was wondering because Input client just stores the JSON as it gets from the server - there's probably lower chance of making mistakes...